src/zenml/logging/step_logging.py

```diff
@@ -252,40 +260,225 @@ def __init__(
                 automatically.
             merge_files_interval: the amount of seconds before the created files
                 get merged into a single file.
+            max_queue_size: maximum number of log batches to queue.
+            queue_timeout: timeout in seconds for putting items in queue when full.
+                - Positive value: Wait N seconds, then drop logs if queue still full
+                - Negative value: Block indefinitely until queue has space (never drop logs)
         """
         # Parameters
         self.logs_uri = logs_uri
         self.max_messages = max_messages
         self.time_interval = time_interval
         self.merge_files_interval = merge_files_interval
+        self.max_queue_size = max_queue_size
+        self.queue_timeout = queue_timeout
 
         # State
         self.buffer: List[str] = []
-        self.disabled_buffer: List[str] = []
         self.last_save_time = time.time()
-        self.disabled = False
         self.artifact_store = artifact_store
 
         # Immutable filesystems state
         self.last_merge_time = time.time()
 
+        # Queue and log storage thread for async processing
+        self.log_queue: queue.Queue[List[str]] = queue.Queue(
+            maxsize=max_queue_size
+        )
+        self.log_storage_thread: Optional[threading.Thread] = None
+        self.shutdown_event = threading.Event()
+        self.shutdown_requested = (
+            False  # Flag to prevent new queue items during shutdown
+        )
+        self.thread_exception: Optional[Exception] = None
+
+        # Buffer lock for thread safety
+        self.buffer_lock = threading.RLock()
+
+        # Start the log storage thread
+        self._start_log_storage_thread()
+
+    def _start_log_storage_thread(self) -> None:
+        """Start the log storage thread for processing log queue."""
+        if (
+            self.log_storage_thread is None
+            or not self.log_storage_thread.is_alive()
+        ):
+            self.log_storage_thread = threading.Thread(
+                target=self._log_storage_worker,
+                name="LogsStorage-Worker",
+            )
+            self.log_storage_thread.start()
+
+    def _log_storage_worker(self) -> None:
+        """Log storage thread worker that processes the log queue."""
+        while not self.shutdown_event.is_set():
+            all_buffers = []
+            try:
+                # Wait for first item in queue or timeout to check shutdown
+                try:
```

Comment:
Do we need to use the buffer_lock here to avoid any race conditions?

Reply:
I don't think so. All the buffers here are already committed to the queue and will not change. Since there is a single thread consuming from the queue, I don't think we need a lock here. Maybe I am missing something?

Reply:
I guess there might be some small inconsistencies depending on when threads try to put stuff in the queue, e.g.
- the log thread starts fetching things from the queue
- some other threads puts stuff in the queue
- without locks, that item now might or might not get picked up by the log thread

Probably not an issue though